### PR TITLE
Fix handling of ranges beyond end of object

### PR DESCRIPTION
**Pull request by adamfaulkner-at**
S3 allows fetching a range that exceeds the end of the object, and returns all of the data that exists within the object.

### CODE REVIEW

**Review by daphnenhuch-at — APPROVED** [review]

File: services/s3/types.go [hunk 1]
```diff
@@ -316,3 +316,13 @@ type ListObjectsV2Output struct {
 	Prefix                *string
 	StartAfter            *string
 }
+
+type InvalidRangeError struct {
```
Comment (AdarLieberDembo-at) [inline services/s3/types.go:320]: Does this need to live in awserrors?

File: services/s3/s3.go [hunk 2]
```diff
@@ -286,10 +295,32 @@ func (s *S3) readerForRange(object *Object, br ByteRange) (io.Reader, *awserrors
...
 
 		readers = append(readers, io.NewSectionReader(f, bytesUntilStart, bytesToReadFromThisChunk))
+		bytesUntilEnd -= (bytesUntilStart + bytesToReadFromThisChunk)
 		bytesUntilStart = 0
-		bytesUntilEnd -= int64(bytesToReadFromThisChunk)
+		readLength += bytesToReadFromThisChunk
+	}
+	if bytesUntilStart > 0 {
+		// For whatever reason, the AWS sdk requires that this error message is formatted correctly.
+		errorObj := InvalidRangeError{
```
Comment (dzbarsky) [inline services/s3/s3.go:304]: lolol, this seems like some poor AWS engineer did a weird thing and then they baked it into the protocol :)

File: services/s3/types.go (hunk 1 above)
Comment (dzbarsky) [reply services/s3/types.go:320]: I think this is fine, this seems like an s3-specific quirk

**Review by dzbarsky — APPROVED** [review]
thanks for writing tests!

**Pull request merged**